# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

In [3]:
import os
os.listdir()

['.ipynb_checkpoints',
 'hw_3_1_Введення_у_Pandas_та_NumPy_.ipynb',
 'hw_3_1_Введення_у_Pandas_та_NumPy_Solution.ipynb',
 'hw_3_2_Основи_роботи_з_даними_у_Pandas.ipynb',
 'hw_3_2_Основи_роботи_з_даними_у_Pandas_Solution.ipynb',
 'hw_3_3_Розширені_методи_обробки_даних_у_Pandas_.ipynb']

In [4]:
df = pd.read_csv('../data/supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [5]:
type(df)

pandas.core.frame.DataFrame

In [6]:
df.head()


Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [8]:
electronic_invoices = df[df['Product line'] == 'Electronic accessories']
electronic_invoices

Invoice ID Branch       City Customer type  Gender  \
1    226-31-3081      C  Naypyitaw        Normal  Female   
5    699-14-3026      C  Naypyitaw        Normal    Male   
6    355-53-5943      A     Yangon        Member  Female   
11   529-56-3974      B   Mandalay        Member    Male   
12   365-64-0515      A     Yangon        Normal  Female   
..           ...    ...        ...           ...     ...   
978  156-20-0370      B   Mandalay        Normal  Female   
984  189-40-5216      C  Naypyitaw        Normal    Male   
988  267-62-7380      C  Naypyitaw        Member    Male   
992  745-74-0715      A     Yangon        Normal    Male   
994  652-49-6720      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
1    Electronic accessories       15.28         5   3.8200   80.2200   
5    Electronic accessories       85.39         7  29.8865  627.6165   
6    Electronic accessories       68.84         6  20.6520  433.6920   
11   Electronic accessories       25.51         4   5.1020  107.1420   
12   Electronic accessories       46.95         5  11.7375  246.4875   
..                      ...         ...       ...      ...       ...   
978  Electronic accessories       25.45         1   1.2725   26.7225   
984  Electronic accessories       96.37         7  33.7295  708.3195   
988  Electronic accessories       82.34        10  41.1700  864.5700   
992  Electronic accessories       58.03         2   5.8030  121.8630   
994  Electronic accessories       60.95         1   3.0475   63.9975   

          Date   Time      Payment    cogs  gross margin percentage  \
1     3/8/2019  10:29         Cash   76.40                 4.761905   
5    3/25/2019  18:30      Ewallet  597.73                 4.761905   
6    2/25/2019  14:36      Ewallet  413.04                 4.761905   
11    3/9/2019  17:03         Cash  102.04                 4.761905   
12   2/12/2019  10:25      Ewallet  234.75                 4.761905   
..         ...    ...          ...     ...                      ...   
978  3/10/2019  18:10  Credit card   25.45                 4.761905   
984   1/9/2019  11:40         Cash  674.59                 4.761905   
988  3/29/2019  19:12      Ewallet  823.40                 4.761905   
992  3/10/2019  20:46      Ewallet  116.06                 4.761905   
994  2/18/2019  11:40      Ewallet   60.95                 4.761905   

     gross income  Rating  
1          3.8200     9.6  
5         29.8865     4.1  
6         20.6520     5.8  
11         5.1020     6.8  
12        11.7375     7.1  
..            ...     ...  
978        1.2725     5.1  
984       33.7295     6.0  
988       41.1700     4.3  
992        5.8030     8.8  
994        3.0475     5.9  

[170 rows x 17 columns]

In [9]:
max_quantity = electronic_invoices['Quantity'].max()
max_price = electronic_invoices['Unit price'].max()
max_quantity, max_price

(10, 99.73)

In [12]:
result = electronic_invoices[(electronic_invoices['Quantity'] == max_quantity) & (electronic_invoices['Unit price'] == max_price)]
result.head()

Empty DataFrame
Columns: [Invoice ID, Branch, City, Customer type, Gender, Product line, Unit price, Quantity, Tax 5%, Total, Date, Time, Payment, cogs, gross margin percentage, gross income, Rating]
Index: []

In [17]:
result = electronic_invoices[(electronic_invoices['Quantity'] == max_quantity)]
result.head()

Invoice ID Branch       City Customer type  Gender  \
75   732-94-0499      C  Naypyitaw        Normal  Female   
105  704-48-3927      A     Yangon        Member    Male   
109  861-77-0145      C  Naypyitaw        Member    Male   
290  139-20-0155      B   Mandalay        Member    Male   
314  227-07-4446      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity  Tax 5%    Total       Date  \
75   Electronic accessories       41.65        10  20.825  437.325  1/13/2019   
105  Electronic accessories       88.67        10  44.335  931.035  1/12/2019   
109  Electronic accessories       81.97        10  40.985  860.685   3/3/2019   
290  Electronic accessories       40.30        10  20.150  423.150  1/24/2019   
314  Electronic accessories       78.13        10  39.065  820.365  2/10/2019   

      Time      Payment   cogs  gross margin percentage  gross income  Rating  
75   17:04  Credit card  416.5                 4.761905        20.825     5.4  
105  14:50      Ewallet  886.7                 4.761905        44.335     7.3  
109  14:30         Cash  819.7                 4.761905        40.985     9.2  
290  17:37  Credit card  403.0                 4.761905        20.150     7.0  
314  20:51         Cash  781.3                 4.761905        39.065     4.4

In [19]:
result = electronic_invoices[(electronic_invoices['Unit price'] == max_price)]
result.head()

Invoice ID Branch      City Customer type  Gender  \
209  817-69-8206      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
209  Electronic accessories       99.73         9  44.8785  942.4485   

         Date   Time      Payment    cogs  gross margin percentage  \
209  3/2/2019  19:42  Credit card  897.57                 4.761905   

     gross income  Rating  
209       44.8785     6.5

Такий рядок не існує, бо рядок з найвищою ціною не має найвищої кількості, а рядок з найвищою кількістю не має у собі найвищої ціни.
Можна сказати, що рядок 209, який містить у собі найвищу ціну по всіх даних, має найвищу кількість у розмірі 9, що не є найвищою кількістю для усіх даних з нашого дата фрейму, а є найвищою кількістю для найвищої ціни. 

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [24]:
def category_mapping(val):
    if val >= 8:
        return 'High'
    elif 6 <= val < 8:
        return 'Medium'
    else:
        return 'Low'

In [26]:
df['Rating Category'] = df.Rating.apply(category_mapping)
df[['Rating', 'Rating Category']].head()

Rating Rating Category
0     9.1            High
1     9.6            High
2     7.4          Medium
3     8.4            High
4     5.3             Low

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [29]:
rating_counts = df['Rating Category'].value_counts()
rating_counts

Rating Category
Medium    356
High      329
Low       315
Name: count, dtype: int64

In [32]:
total_count = df['Rating Category'].count()
total_count

1000

In [34]:
percentage = (rating_counts / total_count) * 100
percentage

Rating Category
Medium    35.6
High      32.9
Low       31.5
Name: count, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [ ]:
time = '11:28'

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.